skrifa/src/outline/unscaled.rs

```diff
@@ -0,0 +1,437 @@
+//! Compact representation of an unscaled, unhinted outline.
+
+#![allow(dead_code)]
+
+use super::DrawError;
+use alloc::vec::Vec;
+use core::ops::Range;
+use raw::{
+    tables::glyf::PointFlags,
+    types::{F26Dot6, Point},
+};
+
+#[derive(Copy, Clone, Default, Debug)]
+pub(super) struct UnscaledPoint {
+    pub x: i16,
+    pub y: i16,
+    pub flags: u16,
+}
+
+impl UnscaledPoint {
+    pub const ON_CURVE: u16 = 1;
+    pub const CUBIC: u16 = 2;
+    pub const CONTOUR_START: u16 = 4;
+
+    pub fn from_glyf_point(
+        point: Point<F26Dot6>,
+        point_flags: PointFlags,
+        is_contour_start: bool,
+    ) -> Self {
+        let point = point.map(|x| (x.to_bits() >> 6) as i16);
+        let mut flags = is_contour_start as u16 * Self::CONTOUR_START;
+        if point_flags.is_on_curve() {
+            flags |= Self::ON_CURVE
+        } else if point_flags.is_off_curve_cubic() {
+            flags |= Self::CUBIC
+        };
+        Self {
+            x: point.x,
+            y: point.y,
+            flags,
+        }
+    }
+
+    pub fn is_on_curve(self) -> bool {
+        self.flags & Self::ON_CURVE != 0
+    }
+
+    pub fn is_off_curve_quad(self) -> bool {
+        self.flags & (Self::ON_CURVE | Self::CUBIC) == 0
+    }
+
+    pub fn is_off_curve_cubic(self) -> bool {
+        self.flags & (Self::CUBIC | Self::ON_CURVE) == Self::CUBIC
+    }
+
+    pub fn is_contour_start(self) -> bool {
+        self.flags & Self::CONTOUR_START != 0
+    }
+}
+
+pub(super) trait UnscaledOutlineSink {
+    fn try_reserve(&mut self, additional: usize) -> Result<(), DrawError>;
+    fn push(&mut self, point: UnscaledPoint) -> Result<(), DrawError>;
+    fn extend(&mut self, points: impl IntoIterator<Item = UnscaledPoint>) -> Result<(), DrawError> {
+        for point in points.into_iter() {
+            self.push(point)?;
+        }
+        Ok(())
+    }
+}
+
+// please can I have smallvec?
+pub(super) struct UnscaledOutlineBuf<const INLINE_CAP: usize> {
+    inline: [UnscaledPoint; INLINE_CAP],
+    inline_len: usize,
+    use_heap: bool,
+    heap: Vec<UnscaledPoint>,
+}
+
+impl<const INLINE_CAP: usize> UnscaledOutlineBuf<INLINE_CAP> {
+    pub fn new() -> Self {
+        Self {
+            inline: [Default::default(); INLINE_CAP],
+            inline_len: 0,
+            use_heap: false,
+            heap: vec![],
+        }
+    }
+
+    pub fn push(&mut self, point: UnscaledPoint) {
+        if !self.use_heap {
+            if self.inline_len < INLINE_CAP {
+                self.inline[self.inline_len] = point;
+                self.inline_len += 1;
+                return;
+            } else {
+                self.use_heap = true;
+                self.heap.extend(&self.inline);
+                self.inline_len = 0;
+            }
+        }
+        self.heap.push(point);
+    }
+
+    pub fn clear(&mut self) {
+        self.inline_len = 0;
+        self.heap.clear();
```

Comment:
should this set `self.use_heap = false`? I guess if you've already allocated you might as well reuse that?

Reply:
yeah, that was my thinking.. we've already allocated so might as well reuse it. Per IM, I'm going to extend the functionality of the already existing internal `SmallArray` type to cover these cases so I can delete this code.